private/bufpkg/bufplugin/bufplugindocker/docker_test.go

```diff
@@ -119,3 +204,136 @@ func buildDockerPlugin(t testing.TB, dockerClient bufplugindocker.Client, docker
 	}
 	return response, err
 }
+
+// dockerServer allows testing some failure flows by simulating the responses to Docker CLI commands.
+type dockerServer struct {
+	httpServer *httptest.Server
+	t          testing.TB
+	buildErr   error
+	pushErr    error
+	// protects builtImages
+	mu          sync.RWMutex
+	builtImages map[string]*builtImage
+}
+
+type builtImage struct {
+	tags []string
+}
+
+func newDockerServer(t testing.TB, version string) *dockerServer {
+	dockerServer := &dockerServer{t: t, builtImages: make(map[string]*builtImage)}
+	versionPrefix := "/v" + version
+	mux := http.NewServeMux()
+	mux.HandleFunc("/session", dockerServer.sessionHandler)
+	mux.HandleFunc(versionPrefix+"/build", dockerServer.buildHandler)
+	mux.HandleFunc(versionPrefix+"/images/", dockerServer.imagesHandler)
+	dockerServer.httpServer = httptest.NewServer(h2c.NewHandler(mux, &http2.Server{}))
+	t.Cleanup(dockerServer.httpServer.Close)
+	return dockerServer
+}
+
+func (d *dockerServer) sessionHandler(w http.ResponseWriter, r *http.Request) {
+	if _, err := io.Copy(io.Discard, r.Body); err != nil {
+		d.t.Error("failed to discard body:", err)
+	}
+	if r.Method != http.MethodPost {
+		http.NotFound(w, r)
+		return
+	}
+	if strings.EqualFold(r.Header.Get("Connection"), "upgrade") && r.ProtoMajor < 2 {
+		w.WriteHeader(http.StatusSwitchingProtocols)
```

Comment:
This is needed to allow h2c upgrades to proceed.

Reply:
Nice, let's write that down too.